### PR TITLE
Checkout

### PR DESCRIPTION
**Pull request by krossovok83**
Step 10
Checkout

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ Metrics/BlockLength:
     - 'app/admin/*.rb'
     - 'db/**/*'
     - 'config/**/*'
+    - 'app/models/concerns/order_state.rb'
```
Comment (YaroslavZahoruiko) [inline .rubocop.yml:26]: please turn off rubocop in file
don't do it in the global config

File: app/models/concerns/order_state.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    aasm column: 'state', whiny_transitions: false do
+      state :created, initial: true
+      state :filled
+      state :address_confirmed
+      state :delivery_confirmed
+      state :payment_confirmed
+      state :order_confirmed
+      state :delivered
+      state :canceled
```
Comment (YaroslavZahoruiko) [inline app/models/concerns/order_state.rb:16]: i think you need
cart, address, delivery, payment, confirmation, delivered, cancelled

File: app/models/concerns/order_state.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+      event :come_back do
```
Comment (YaroslavZahoruiko) [inline app/models/concerns/order_state.rb:18]: pass_cart_step

File: app/models/concerns/order_state.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        transitions from: %i[filled address_confirmed delivery_confirmed payment_confirmed], to: :created
+      end
+
+      event :filling do
```
Comment (YaroslavZahoruiko) [inline app/models/concerns/order_state.rb:22]: pass_ address_step

File: app/javascript/views/checkout_address.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+
+// const form = document.getElementById("shipping");
+// const shipping_collect = form ? Array.prototype.slice.call(form.elements) : [];
+// const checkbox = document.getElementById("checkbox_shipping")
+//
+// checkbox && checkbox.addEventListener("change", function ChekBoxShipping() {
+//     shipping_collect.forEach((elem) => elem.readOnly = true);
+// });
```
Comment (YaroslavZahoruiko) [inline app/javascript/views/checkout_address.js:8]: remove

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -9,4 +9,8 @@ def flash_class(level)
     when :alert then Constants::ALERT_DANGER
     end
   end
+
+  def euro(number)
+    number_to_currency(number, unit: '€')
+  end
```
Comment (YaroslavZahoruiko) [inline app/helpers/application_helper.rb:15]: please do not use helpers

File: app/controllers/checkouts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  def update_step
+    case step
+    when :addresses then CheckoutServices::UpdateAddresses.call(current_order, params)
+    when :delivery then CheckoutServices::UpdateDelivery.call(current_order, params)
+    when :payment then CheckoutServices::UpdatePayment.call(current_order, params)
+    when :confirm then confirm_order
+    else
+      head(:bad_request) && return
+    end
+  end
```
Comment (YaroslavZahoruiko) [inline app/controllers/checkouts_controller.rb:36]: please move all business logic to service object

File: app/controllers/checkouts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  steps :addresses, :delivery, :payment, :confirm, :complete
+
+  def show
+    show_complete if current_step?(:complete)
+    render_wizard
+  end
+
+  def edit
+    render_wizard
+  end
```
Comment (YaroslavZahoruiko) [inline app/controllers/checkouts_controller.rb:19]: do you need edit action?

File: app/services/order_services/current_order.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+module OrderServices
+  class CurrentOrder < ApplicationService
```
Comment (YaroslavZahoruiko) [inline app/services/order_services/current_order.rb:4]: This class is called current_order but in this class you are initializing the addresses, it violates the Single Responsibility Principle

**Review by YaroslavZahoruiko — CHANGES_REQUESTED** [review]

File: app/models/concerns/order_state.rb (hunk 2 above)
Comment (krossovok83) [reply app/models/concerns/order_state.rb:16]: if the question is about naming, then corrected
if to quantity:
created - cart - very convenient to validate the presence of orders in the cart
in_delivery - according to the task, it must be present

File: app/services/checkout_services/check_state.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    end
+
+    def call
+      case @step
+      when :addresses then @order.cart?
+      when :delivery then @order.address?
+      when :payment then @order.delivery?
+      when :confirm then @order.payment?
+      when :complete then @order.confirmation?
+      end
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/check_state.rb:17]: perhaps we need to check for other parameters such as the presence of an address or order_items

File: app/services/checkout_services/order_confirm.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    def call
+      order_item_price
+      @order.attributes = new_attributes
+      @order.save if @order.pass_confirmation_step!
+    end
+
+    private
+
+    def order_number
+      "R#{8.times.map { rand(10) }.join}"
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/order_confirm.rb:20]: often use the creation date of the order for its number, and the code looks nice
"R#{Time.zone.now.to_s(:number)}"

File: app/services/checkout_services/update_addresses.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    def update_params
+      {
+        billing_address_attributes: @params.require(:billing_address_attributes).permit(address_params),
+        shipping_address_attributes: @params.require(shipping_address_key).permit(address_params)
+      }
+    end
+
+    def shipping_address_key
+      @params[:use_billing] == '1' ? :billing_address_attributes : :shipping_address_attributes
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/update_addresses.rb:31]: what is '1' ?

File: app/services/checkout_services/update_step.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      when :confirm then confirm_order
+      else false
+      end
+    end
+
+    private
+
+    def confirm_order
+      CheckoutServices::OrderConfirm.call(@user, @order, @summary) &&
+        (@session = 'delete')
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/update_step.rb:29]: remove session you can stay in controller
session[:order_id] = nil if order.confirmed?

File: app/services/order_services/current_order.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    end
+
+    private
+
+    def find_order
+      @order = Order.find_by(token: @token)
+    end
+
+    def build_addresses
+      OrderServices::BuildAddresses.call(@order, @user) if @user
```
Comment (YaroslavZahoruiko) [inline app/services/order_services/current_order.rb:23]: this is not necessary, put one line of code into a separate method

File: app/services/order_services/current_order.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    def call
+      find_order
+      build_addresses
+      @order
+    end
+
+    private
+
+    def find_order
+      @order = Order.find_by(token: @token)
```
Comment (YaroslavZahoruiko) [inline app/services/order_services/current_order.rb:19]: the same

Comment (YaroslavZahoruiko) [reply app/services/order_services/current_order.rb:19]: you can do that in TODO list in the trailblazer

**Review by YaroslavZahoruiko — CHANGES_REQUESTED** [review]

File: app/models/concerns/order_state.rb [hunk 16]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module OrderState
+  extend ActiveSupport::Concern
+  # rubocop:disable Metrics
+  included do
+    include AASM
+
+    aasm column: 'state', whiny_transitions: false do
+      state :created, initial: true
```
[the remark this answers is not in the document]
Comment (krossovok83) [reply app/models/concerns/order_state.rb:10]: for validation on a non-empty cart

File: app/controllers/checkouts_controller.rb [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+  def update
+    result = CheckoutServices::UpdateStep.new(step, current_order, params, current_user)
+    if result.call
+      session.delete(:token) if current_order.confirmation?
+      redirect_to_next_step
+    else
+      redirect_to_next step
```
Comment (YaroslavZahoruiko) [inline app/controllers/checkouts_controller.rb:22]: why, when your services fail, are you going to take the next step?

File: app/models/order.rb [hunk 18]
```diff
@@ -1,6 +1,22 @@
...
+  accepts_nested_attributes_for :billing_address, :shipping_address
   has_one :coupon
+  belongs_to :delivery, optional: true
+  has_one :card, dependent: :destroy
+  accepts_nested_attributes_for :card
+
+  validates_presence_of :order_items, if: proc { cart? }, message: 'You cart is empty'
+  validates_presence_of :billing_address, :shipping_address, if: proc { address? },
+                                                             message: 'Two addresses are required for ordering'
+  validates :delivery_id, presence: true, if: proc { delivery? }
```
Comment (YaroslavZahoruiko) [inline app/models/order.rb:19]: i18n

File: app/services/checkout_services/check_state.rb [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+      @order = order
+      @step = step
+    end
+
+    def call
+      return true if @step.in?(%i[addresses delivery payment confirm]) && @order.payment?
+
+      case @step
+      when :addresses then @order.cart?
+      when :delivery then @order.state == 'address'
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/check_state.rb:15]: constant

File: app/controllers/checkouts_controller.rb [hunk 20]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+class CheckoutsController < ApplicationController
+  include Wicked::Wizard
+  before_action :authenticate_user!
+
+  steps :addresses, :delivery, :payment, :confirm, :complete
```
Comment (YaroslavZahoruiko) [inline app/controllers/checkouts_controller.rb:7]: STEPS = { addresses: 'addresses' }
steps STEPS[:addresses]

File: app/services/checkout_services/check_state.rb (hunk 19 above)
Comment (YaroslavZahoruiko) [reply app/services/checkout_services/check_state.rb:15]: STEPS[:addresses]

File: app/services/checkout_services/check_state.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+      when :payment then payment
+      when :complete then @order.confirmation?
+      end
+    end
+
+    private
+
+    def payment
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/check_state.rb:23]: your step is payment, why check that the order is in the delivery status?

File: app/services/checkout_services/card_define.rb [hunk 22]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module CheckoutServices
+  class CardDefine < ApplicationService
+    def initialize(order)
+      @order = order
+    end
+
+    def call
+      @order.build_card unless @order.card
```
Comment (YaroslavZahoruiko) [inline app/services/checkout_services/card_define.rb:10]: i think you do not need service object for initialize card without any conditions

**Review by YaroslavZahoruiko — CHANGES_REQUESTED** [review]

File: app/controllers/checkouts_controller.rb [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+  before_action :authenticate_user!
+
+  steps(*Constants::STEPS.values)
+
+  def show
+    @order_summary = OrderPresenters::Summary.new(current_order)
+    case step
+    when Constants::STEPS[:payment] then current_order.build_card if current_order.card.nil?
+    when Constants::STEPS[:complete] then show_complete
+    end
```
Comment (YaroslavZahoruiko) [inline app/controllers/checkouts_controller.rb:14]: move to service object

File: app/controllers/orders_controller.rb [hunk 24]
```diff
@@ -1,19 +1,33 @@
 # frozen_string_literal: true
 
 class OrdersController < ApplicationController
+  before_action :to_cart_state
```
Comment (YaroslavZahoruiko) [inline app/controllers/orders_controller.rb:4]: move to service object

File: app/controllers/orders_controller.rb [hunk 25]
```diff
@@ -1,19 +1,33 @@
...
   def show
     @order_items = current_order.order_items.includes(:book)
     @sum = OrderPresenters::Summary.new(current_order)
   end
 
   def update
-    result = OrderServices::CouponChecker.new(current_order, params)
+    result = OrderServices::ToCheckout.new(current_order)
```
Comment (YaroslavZahoruiko) [inline app/controllers/orders_controller.rb:11]: what do you mean by ToCheckout? it's a bad name for a service

**Review by YaroslavZahoruiko — CHANGES_REQUESTED** [review]

**Review by YaroslavZahoruiko — APPROVED** [review]

**Pull request merged**